game/procurement.py

```diff
@@ -182,26 +182,49 @@ def most_needed_unit_class(self, cp: ControlPoint) -> GroundUnitClass:
     @staticmethod
     def fulfill_aircraft_request(
         squadrons: list[Squadron], quantity: int, budget: float
-    ) -> Tuple[float, bool]:
+    ) -> Tuple[float, bool, bool]:
+        """
+        Purchases aircraft for a requested task. Arguments:
+            List of potential squadrons to consider the task.
+            Requested quantity.
+            Available budget.
+        Returns:
+            A tuple containing, respectively, a float (remaining budget), a bool (true if the request was fulfilled)
+            and a bool (false if the request was not fulfilled because of insufficient pilots, prompting a retry).
+        """
         for squadron in squadrons:
+            insufficient_pilots = False
             price = squadron.aircraft.price * quantity
+
+            # Final check to make sure the number of aircraft won't exceed the number of available pilots
+            # after fulfilling this aircraft request.
+            if (
+                squadron.pilot_limits_enabled
+                and squadron.expected_size_next_turn + quantity > squadron.max_size
+            ):
+                insufficient_pilots = True
+                continue
             if price > budget:
                 continue
 
             squadron.pending_deliveries += quantity
             budget -= price
-            return budget, True
-        return budget, False
+            return budget, True, insufficient_pilots
+        return budget, False, insufficient_pilots
 
     def purchase_aircraft(self, budget: float) -> float:
         for request in self.game.coalition_for(self.is_player).procurement_requests:
             squadrons = list(self.best_squadrons_for(request))
             if not squadrons:
                 # No airbases in range of this request. Skip it.
                 continue
-            budget, fulfilled = self.fulfill_aircraft_request(
+            budget, fulfilled, insufficient_pilots = self.fulfill_aircraft_request(
                 squadrons, request.number, budget
             )
+            if insufficient_pilots:
+                # The request was not fulfilled because there were not enough pilots in any of the provided squadrons
+                # and not because of insufficient budget. Continue processing the aircraft requests.
+                continue
```

Comment:
Won't this cause a similar problem as described on L229? These requests are in a priority order, and if we can't fulfill an early item then we probably can't plan the mission the later items were intended for either way.

If the orders here are something like:

2xDEAD
2xSEAD Escort
2xEscort
2xDEAD
2xSEAD Escort
2xEscort
2xDEAD
2xSEAD Escort
2xEscort

If there are only 6 parking spaces available, plenty of budget, and plenty of pilots for escort missions but none for SEAD Escort/DEAD, you'll end up filling the airport with escorts that can never be used because there will never be fulfilled DEAD missions for them to escort.

Reply:
That is a possibility. I haven't encountered such cases yet in any of my campaigns, but I tend to design my own campaigns and my squadron configuration might work around the problem.

Escorts can (in most cases) be used for BARCAP, so the situation might resolve itself when some of them get shot down?

I think there needs to be some way to distinguish between purchase orders scrapped because of insufficient funds and orders scrapped because of insufficient pilots. Otherwise the AI will almost stop buying aircraft altogether when squadrons get low on pilots.

Reply:
One possible solution would be to use hierachy in the aircraft procurement requests. Instead of putting the escorts in a flat list with the SEAD/DEAD, the escorts would be included in the same request. The we could scrub the whole mission including the escorts in case one of the squadrons runs out of pilots?

Reply:
It's not something you see for bluefor because most aircraft are multirole. It absolutely wrecks red factions because they end up buying hundreds of MiG-19s and nothing else and can then never plan any missions.

Purchases are in the order that missions are planned, and missions are often dependent on earlier missions. If you miss the first purchase but manage to fill every other purchase, there's no guarantee that _any_ additional missions will be planned.

I don't think there's a meaningful difference in why the purchase failed. Without explicit dependency tracking for packages and purchase orders, we cannot continue after an unfulfilled purchase regardless of why it happened or we'll cause the problem described by that comment.

Reply:
How about I make it an option (default: off) in the autoplanner settings under Campaign Management? By default, the first scrubbed purchase would cancel all the rest and the user could separately enable a setting to have the autoplanner continue processing subsequent orders if a squadron with insuffient pilots is encountered.

Reply:
I don't think a "buy aircraft that can't be used and potentially fill all the parking so it breaks my game" checkbox is something we should add. This isn't a theoretical problem. This is why a ton of factions couldn't sortie any missions but CAP several releases ago. Players will not notice that kind of failure when it happens to red because the only symptom is that the game is boring and easy.